### PR TITLE
Alternative sobol reducer using higher-order functions

### PR DESCRIPTION
**Pull request by melsman**
Maybe we should rename the function to something else than `redomap`?

### CODE REVIEW

**Comment by melsman** [comment]
Sobol-sequences are quite nice:

image

**Comment by Athas** [comment]
But `pi` converges immediately!  Why not just use that?

**Comment by melsman** [comment]
Good question... Maybe sometimes you don't know what pi is! ;)

fre. 25. maj 2018 kl. 16.39 skrev Troels Henriksen <notifications@github.com

File: futlib/sobol.fut [hunk 1]
```diff
@@ -26,6 +26,7 @@ module type sobol = {
   module Reduce :
       (X : { include monoid
              val f : [D]f64 -> t }) -> { val run : i32 -> X.t }
+  val redomap 't : ([D]f64 -> t) -> (t -> t -> t) -> t -> i32 -> t
```
Comment (Athas) [inline futlib/sobol.fut]: Why is the neutral element required?  Could we not simply require that if the chunk function is applied to an empty array of `f64`s, then the result must be the neutral element?

**Comment by Athas** [comment]
Why do we use `f64` in the interface?  As far as I can see, we construct them from 32-bit integers anyway.

**Comment by melsman** [comment]
... it's not a chunk... it's an element in the D-dimensional space...

fre. 25. maj 2018 kl. 16.42 skrev Troels Henriksen <notifications@github.com

**Comment by melsman** [comment]
Good point. I guess there are two alternatives:

1. Use u32 as the type. In this case the use will need to do some manual
conversions; I don't think I like that...

2. Somehow parameterise the module over the type...

fre. 25. maj 2018 kl. 16.42 skrev Troels Henriksen <notifications@github.com

**Comment by Athas** [comment]
Instead of `redomap`, why not `run` or `simulate`?

Also, the test looks like it's failing.

**Comment by melsman** [comment]
don't merge yet...

**Comment by melsman** [comment]
yes - if tiu think it's fine ;)

søn. 27. maj 2018 kl. 12.22 skrev Troels Henriksen <notifications@github.com

**Pull request merged**